pkg/verification/verify.go

```diff
@@ -18,17 +18,196 @@ package verification
+)
+
+// VerifyOpts contains verification options passed via the CLI vefify command
+// These fields are then used to verify the TSR
+type VerifyOpts struct {
+	// verifies that the TSR's OID has an expected value
+	OID asn1.ObjectIdentifier
+	// verifies that the TSR uses the TSACertificate as expected
+	TSACertificate *x509.Certificate
+	// verifies the TSR's certificate chain with the root certificates
+	Intermediates []*x509.Certificate
+	// verifies the TSR's certificate chain with the intermediate certificates
+	Roots []*x509.Certificate
+	// verifies that the TSR contains the expected nonce that was optionally
+	// passed to the TSA when requesting a timestamp
+	Nonce *big.Int
+	// verifies that the leaf certificate subject Common Name an expected value
+	CommonName string
+}
+
+// Verify the TSR's certificate identifier matches a provided TSA certificate
+func verifyESSCertID(tsaCert *x509.Certificate, opts VerifyOpts) error {
+	if opts.TSACertificate == nil {
+		return nil
+	}
+
+	if !bytes.Equal(opts.TSACertificate.RawIssuer, tsaCert.RawIssuer) {
+		return fmt.Errorf("TSR cert issuer does not match provided TSA cert issuer")
+	}
+
+	if opts.TSACertificate.SerialNumber.Cmp(tsaCert.SerialNumber) != 0 {
+		return fmt.Errorf("TSR cert serial number does not match provided TSA cert serial number")
+	}
+
+	return nil
+}
+
+// Verify the leaf certificate's subject Common Name name matches a provided Common Name
+func verifySubjectCommonName(cert *x509.Certificate, opts VerifyOpts) error {
+	if opts.CommonName == "" {
+		return nil
+	}
+
+	if cert.Subject.CommonName != opts.CommonName {
+		return fmt.Errorf("the certificate's subject Common Name %s does not match the provided Common Name %s", cert.Subject.CommonName, opts.CommonName)
+	}
+	return nil
+}
+
+// If embedded in the TSR, verify the TSR's leaf certificate matches a provided TSA certificate
+func verifyEmbeddedLeafCert(tsaCert *x509.Certificate, opts VerifyOpts) error {
+	if opts.TSACertificate != nil && !opts.TSACertificate.Equal(tsaCert) {
+		return fmt.Errorf("certificate embedded in the TSR does not match the provided TSA certificate")
+	}
+	return nil
+}
+
+// Verify the leaf's EKU is set to critical, per RFC 3161 2.3
+func verifyLeafCertEKU(cert *x509.Certificate) error {
+	var criticalEKU bool
+	for _, ext := range cert.Extensions {
+		if ext.Id.Equal(EKUOID) {
+			criticalEKU = ext.Critical
+		}
+	}
+	if !criticalEKU {
+		return errors.New("certificate must set EKU to critical")
+	}
+	return nil
+}
+
+func verifyLeafCert(ts timestamp.Timestamp, opts VerifyOpts) error {
+	if len(ts.Certificates) == 0 && opts.TSACertificate == nil {
+		return fmt.Errorf("leaf certificate must be present the in TSR or as a verify option")
+	}
+
+	errMsg := "failed to verify leaf cert"
+
+	var leafCert *x509.Certificate
+	if len(ts.Certificates) != 0 {
+		leafCert = ts.Certificates[0]
+		if opts.TSACertificate != nil && !leafCert.Equal(opts.TSACertificate) {
```

Comment:
Is this the same check you do in `verifyEmbeddedLeafCert`?

Reply:
Yes, it is. I'll drop this check.